File: JavaScript/0806_number_of_lines_to_write_string/solution.js
function number_of_lines_to_write_string(widths,string) {
    let refer = {};
    let width_index = 0;
    for (let i = 'a'.charCodeAt(); i <= 'z'.charCodeAt(); i ++) {
        if (!(String.fromCharCode(i) in refer)) {
            refer[String.fromCharCode(i)] = widths[width_index];
        }
        width_index ++;
    }
    let final = [];
    let result = [];
    for (let letter of string) {
        result.push(refer[letter])
        const total = result.reduce((acc, val) => acc + val, 0);
        if (total > 100) {
            result.pop();
            final.push(total);
            result = [];
            result.push(refer[letter]);
        }
    }
    const total = result.reduce((acc, val) => acc + val, 0);
    final.push(total);
    return [final.length,final[final.length - 1]];
}

const widths = [10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10];
const string = "abcdefghijklmnopqrstuvwxyz";
console.log(number_of_lines_to_write_string(widths,string));

File: Python/0806_number_of_lines_to_write_string/solution.py
def number_of_lines_to_write_string(widths,string):  
    refer = dict()
    width_index = 0
    for index in range(ord('a'),ord('z') + 1):
        if chr(index) not in refer:
            refer[chr(index)] = widths[width_index]
        width_index += 1
    final = list()
    result = list()
    for letter in string:
        result.append(refer[letter])
        if sum(result) > 100:
            result.pop()
            final.append(sum(result))
            result.clear()
            result.append(refer[letter])
    final.append(sum(result))
    return [len(final),final[-1]]
            

widths = [4,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10]
string = "bbbcccdddaaa"
print(number_of_lines_to_write_string(widths,string))
